File: backend/utils/jwt.js
const jwt = require('jsonwebtoken');
const { PrismaClient } = require('../prisma/generated/prisma');
const prisma = new PrismaClient()

if (!process.env.JWT_SECRET || !process.env.JWT_EXPIRATION) {
    console.error('JWT_SECRET or JWT_EXPIRATION is not defined in the environment variables');
    process.exit(1);
}

const SECRET_KEY = process.env.JWT_SECRET;
const JWT_EXPIRATION = process.env.JWT_EXPIRATION;

// Generate a JWT token
exports.generateToken = (payload, expiresIn = JWT_EXPIRATION) => {
    return jwt.sign(payload, SECRET_KEY, { expiresIn });
};

// Verify a JWT token manually BUT USE THE AUTH MIDDLEWARE
exports.verifyToken = (token) => {
    try {
        return jwt.verify(token, SECRET_KEY);
    } catch (err) {
        throw new Error('Invalid or expired token');
    }
};

exports.isTokenBlacklisted = async (token) => {
    // Check if jwt token is blacklisted
    const user = await prisma.bljwts.findFirst({
        where: {
            jwt: token
        }
    });
    if (user) {
        return true;
    }
    return false;
}

File: backend/utils/jwt.py
import os
import jwt
from datetime import datetime, timedelta
from ORM.db import get_db
from ORM.models import blJWTs

SECRET_KEY = os.getenv("JWT_SECRET")
JWT_EXPIRATION = os.getenv("JWT_EXPIRATION", "1h")

if not SECRET_KEY or not JWT_EXPIRATION:
    raise RuntimeError("JWT_SECRET or JWT_EXPIRATION is not defined in the environment variables")

def parse_expiration(exp_str):
    if exp_str.endswith('h'):
        hours = int(exp_str[:-1])
        return timedelta(hours=hours)
    raise ValueError("Unsupported expiration format")

def generate_token(payload, expires_in=JWT_EXPIRATION):
    exp_delta = parse_expiration(expires_in)
    payload = payload.copy()
    payload['exp'] = datetime.utcnow() + exp_delta
    return jwt.encode(payload, SECRET_KEY, algorithm="HS256")

def verify_token(token):
    try:
        return jwt.decode(token, SECRET_KEY, algorithms=["HS256"])
    except jwt.ExpiredSignatureError:
        raise Exception("Token expired")
    except jwt.InvalidTokenError:
        raise Exception("Invalid token")

def is_token_blacklisted(token):
    with get_db() as db:
        result = db.query(blJWTs).filter(blJWTs.jwt == token).first()
        return bool(result)
